dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/trackedentity/JdbcTrackedEntityChangeLogStore.java

```diff
@@ -77,38 +106,52 @@ public Page<TrackedEntityChangeLog> getTrackedEntityChangeLog(
               case
           when cl.type = 'DELETE' then cl.currentchangelogvalue
           when cl.type = 'UPDATE' then cl.previouschangelogvalue
-          end as previousvalue, cl.createdat , cl.trackedentityattributeuid, cl.firstname, cl.surname, cl.username, cl.useruid, cl.type
+          end as previousvalue, cl.created, cl.trackedentityattributeuid, cl.firstname, cl.surname, cl.username, cl.useruid, cl.type
           from
               (
-                  select audit.created as createdat, tea.uid as trackedentityattributeuid, audit.audittype as type, u.firstname, u.surname, u.username, u.uid as useruid,
+                  select audit.created, tea.uid as trackedentityattributeuid, audit.audittype as type, u.firstname, u.surname, u.username, u.uid as useruid,
                   lead (audit.value) over (partition by audit.trackedentityid, audit.trackedentityattributeid order by audit.created DESC) as previouschangelogvalue,
                   audit.value as currentchangelogvalue
                   from trackedEntityAttributeValueAudit audit
                   join trackedentity t using(trackedentityid)
                   join trackedentityattribute tea using(trackedentityattributeid)
                   join userinfo u on u.username = audit.modifiedby
-                  where audit.audittype in ('CREATE', 'UPDATE', 'DELETE')
-                  and t.uid = :trackedEntity
        """;
-    MapSqlParameterSource parameters =
-        new MapSqlParameterSource("trackedEntity", trackedEntity.getValue())
-            .addValue("limit", pageParams.getPageSize() + 1)
-            .addValue("offset", (pageParams.getPage() - 1) * pageParams.getPageSize());
+
+    if (program != null) {
+      sql +=
+          """
+              join program_attributes pa using (trackedentityattributeid)
+              join program p using (programid)
+              where audit.audittype in ('CREATE', 'UPDATE', 'DELETE')
+              and t.uid = :trackedEntity
+              and p.uid = :programUid
+          """;
+      parameters.addValue("programUid", program.getValue());
+    } else {
+      sql +=
+          """
+              where audit.audittype in ('CREATE', 'UPDATE', 'DELETE')
+              and t.uid = :trackedEntity
+          """;
+    }
 
     List<TrackedEntityChangeLog> changeLogs;
     if (attributes.isEmpty()) {
       sql +=
           """
-              order by audit.created desc) cl
-              limit :limit offset :offset
-          """;
+              order by %s
+              limit :limit offset :offset) cl
+          """
+              .formatted(sortExpressions(order));
     } else {
       sql +=
           """
               and tea.uid in (:attributes)
```

Comment:
You could even invert the if condition and add the `and tea.uid in (:attributes)` to `sql` then at the end add the limit/offset. This would reduce some nesting.